README.md
# Micro Mockers

[![Build Status](https://travis-ci.org/Wiredcraft/micro-mockers.svg?branch=master)](https://travis-ci.org/Wiredcraft/micro-mockers) [![Coverage Status](https://coveralls.io/repos/github/Wiredcraft/micro-mockers/badge.svg?branch=master)](https://coveralls.io/github/Wiredcraft/micro-mockers?branch=master)

Mock multiple (micro-)services with Docker boxes and gateway with Kong.

## How to use

- _Always shutdown the docker boxes before you change anything_
- Download and install Docker
- `npm install -g micro-mockers`
- `cd` to your work directory (see `test/fixture` as an example)
- `mm build`
- `mm up`
- `mm down`

## Features

### Mock service boxes management

_TODO_

### Kong admin API libraries

```js
const mm = require('micro-mockers');
const adminApi = mm.kong.adminApi;
```

`Status` can be used to ping the Admin API.

```js
const status = new adminApi.Status('http://localhost:8001');
status.ping([max]).then(...);
```

`Plugins` can be used to ensure what in Kong matches an array of plugin definitions.

```js
const plugins = new adminApi.Plugins('http://localhost:8001');
plugins.syncAll([{
  name: 'rate-limiting',
  config: {
    hour: 3000
  }
}, {
  name: 'syslog',
  config: {}
}]).then(...);
```

`Apis` can be used to ensure what in Kong matches an array of API definitions.

```js
const apis = new adminApi.Apis('http://localhost:8001');
apis.syncAll([{
  name: 'lorem',
  hosts: ['...'],
  upstream_url: '...',
  plugins: [...]
}, ...]).then(...);
```


History.md

0.5.0 / 2018-05-08
==================

  * Dropped Node 6.
  * Use eslint.
  * Upgrade to Kong 0.12
  * Removed example - there's already the test/fixture.
  * Upgrade to loopback-boot v3 and use Yarn.

0.4.1 / 2017-08-10
==================

  * Added request lib for Kong consumers API.
  * Added `.unknown(true)` to all `Joi.object()`.

0.4.0 / 2017-07-20
==================

  * Updated readme.
  * Updated to Kong 0.10.

0.3.2 / 2017-01-12
==================

  * Give adminApi a default address.

0.3.1 / 2017-01-11
==================

  * Tiny fix: should be a getter, to keep consistency.

0.3.0 / 2017-01-11
==================

  * Improved Kong admin API classes to have the ability to do sync (add, update, delete) according to the config.
  * Refactored Config class.

0.2.0 / 2016-12-28
==================

  * Better Makefile.
  * Rebuilt the Builder, Config, and Kong API classes.
  * Include all scripts with bin.

0.1.1 / 2016-11-16
==================

  * Use `lib-rest` for Kong admin API. Use Docker for the tests.
  * Cover more relative paths.
  * Small improvements and cleanups.

0.1.0 / 2016-09-28
==================

* First release.


lib/classes/Config.js
'use strict';

const path = require('path');
const dotProp = require('dot-prop');
const merge = require('mixable-object').merge;
const Bootstrapper = require('loopback-boot').Bootstrapper;

const Promise = require('bluebird');
const fs = Promise.promisifyAll(require('fs'));

const lib = require('../');

// Extended Joi.
const Joi = lib.extended.Joi;

// Schema for Kong consumers.
const schemaConsumers = Joi.array().items(Joi.object({
  username: Joi.string().trim(),
  custom_id: Joi.string().trim()
}).unknown(true));
// Schema for Kong plugin configs.
const schemaPlugins = Joi.array().items(Joi.object({
  name: Joi.string().trim(),
  config: Joi.object().unknown(true)
}).unknown(true));
// Schema for Kong API configs.
const schemaApis = Joi.array().items(Joi.object({
  name: Joi.string().trim(),
  hosts: Joi.array().items(Joi.string().trim()),
  uris: Joi.array().items(Joi.string().trim()),
  methods: Joi.array().items(Joi.string().trim()),
  strip_uri: Joi.boolean(),
  preserve_host: Joi.boolean(),
  upstream_url: Joi.string().trim(),
  retries: Joi.number(),
  upstream_connect_timeout: Joi.number(),
  upstream_send_timeout: Joi.number(),
  upstream_read_timeout: Joi.number(),
  https_only: Joi.boolean(),
  http_if_terminated: Joi.boolean(),
  plugins: schemaPlugins
}).unknown(true));

module.exports = class Config {
  constructor(options) {
    if (options == null) {
      options = {};
    } else if (typeof options === 'string') {
      options = { root: options };
    }
    this.root = options.root = options.root || process.cwd();
    this.env = options.env || process.env.NODE_ENV || 'development';
    this._bootstrapper = new Bootstrapper({
      // app: this.env,
      appRootDir: this.root,
      phases: ['load'],
      plugins: ['application']
    });
    this._context = {
      // app: this.env,
    };
    this._load = null;

    /**
     * Resolve path to absolute, assuming it's relative to the root path.
     *
     * Bound to this.
     */
    this.resolvePath = (source) => {
      if (typeof source !== 'string') {
        return source;
      }
      return path.resolve(this.root, source);
    };

    /**
     * Convert the path from "relative to config file" to "relative to compose file".
     *
     * Bound to this.
     */
    this.relativeToCompose = (source) => {
      if (typeof source !== 'string' || !source.startsWith('.')) {
        return source;
      }
      let res = path.relative(this.composeRoot, path.resolve(this.root, source));
      if (!res.startsWith('.')) {
        // TODO: what about Windows?
        res = './' + res;
      }
      return res;
    };
  }

  setup() {
    this.settings = this._context.configurations.application;

    // This is how the config should look like.
    const schema = {
      // For the docker compose file.
      compose: Joi.string().trim().transform(this.resolvePath),
      template: Joi.string().trim().transform(this.resolvePath),
      types: Joi.object({
        mock: Joi.object({
          build: Joi.string().trim().transform(this.relativeToCompose)
        }).unknown(true),
        doc: Joi.object({
          build: Joi.string().trim().transform(this.relativeToCompose)
        }).unknown(true)
      }).unknown(true),
      services: Joi.array().items(Joi.object({
        type: Joi.string().trim(),
        volume: Joi.string().trim().transform(this.relativeToCompose)
      }).unknown(true)),
      // For the Kong config.
      adminApi: Joi.string().trim().default('http://localhost:8001'),
      consumers: schemaConsumers,
      plugins: schemaPlugins,
      apis: schemaApis
    };
    this.schema = Joi.object(schema).unknown(true);

    // Shortcuts.
    for (let key in schema) {
      if (!(key in this)) {
        Object.defineProperty(this, key, {
          get: function() {
            return this.get(key);
          }
        });
      }
    }
    Object.defineProperty(this, 'composeRoot', {
      get: function() {
        return path.dirname(this.get('compose'));
      }
    });
  }

  load(done) {
    // Cached with a promise.
    if (this._load != null) {
      return this._load.asCallback(done);
    }
    this._load = Promise.resolve(this._bootstrapper.run(this._context))
      .bind(this).then(this.setup).return(this);
    return this._load.asCallback(done);
  }

  get(key) {
    const schema = Joi.reach(this.schema, key);
    if (schema == null) {
      return dotProp.get(this.settings, key);
    }
    return Joi.attempt(dotProp.get(this.settings, key), schema);
  }

  set(key, val) {
    const schema = Joi.reach(this.schema, key);
    if (schema == null) {
      return dotProp.set(this.settings, key, val);
    }
    return dotProp.set(this.settings, key, Joi.attempt(val, schema));
  }

  readTemplate() {
    return this.load().then(() => {
      return fs.readFileAsync(this.get('template'), 'utf8');
    });
  }

  /**
   * Shortcut.
   */
  get services() {
    const types = this.get('types');
    const res = this.get('services');
    for (let key in res) {
      let service = res[key];
      if (service.type != null && types[service.type] != null) {
        res[key] = merge.call(Object.assign({}, types[service.type]), res[key]);
      }
    }
    return res;
  }

  /**
   * Shortcut.
   */
  get apis() {
    const res = this.get('apis');
    if (res) {
      return res;
    }
    // Build from services.
    return this.services.map(this.serviceToApi);
  }

  /**
   * Helper.
   */
  serviceToApi(service) {
    const json = {
      name: service.name,
      upstream_url: 'http://' + service.name + ':' + (service.port || '3000')
    };
    if (service.proxyPath) {
      json.uris = [service.proxyPath];
      json.strip_uri = true;
    } else {
      json.hosts = [service.name];
    }
    return json;
  }
};
